classes/helper.php

```diff
@@ -549,6 +550,14 @@ public static function acknowledge_notice($noticeid) {
         } else {
             $result['status'] = false;
         }
+
+        $notice = sitenotice::get_record(['id' => $noticeid]);
+        if ($notice->get('forcelogout')) {
+            require_logout();
+            $loginpage = new \moodle_url("/login/index.php");
+            $result['redirecturl'] = $loginpage->out();
```

Comment:
Would you like to abstract this to a function to reduce duplicated code?

Reply:
Good idea, moved in to a new function called `optionally_logout`

Reply:
Ok. We got rid of the duplicated code, but honestly it became even worse. This method now is confusing and does few things that you would not expect after looking at the name of the method. 

Also it seems like using it inside acknowledge_notice method breaks a logic as even if you acknowledged a notice you still get kicked out as we check for ($notice->get('forcelogout') || $notice->get('reqack')). 

I'm wondering if it's possible to cover the logic with tests to make sure we don't break it with the changes?

Reply:
Good catch with the `acknowledge_notice` thing. I've updated that method so it now has an `acknowledged` param to prevent that from happening.

Do you have any suggestions for a better name or how to make it less confusing? I'm wondering if it's clearer to just move the logic back in to the `acknowledge_notice` and `dismiss_notice` methods.

Can investigate unit tests once we decide exactly where the best spot for this code is.

Reply:
yeah I guess moving  the logic back would be a best option. It feels like the code needs a good refactoring. Not as part of this issue. However, having written tests will be really helpful.

Reply:
I've removed the function `optionally_logout` and left the logic in the individual functions. I've created #46 to capture the refactor work in the future.